### DISCUSSION TITLE
getState() not grabbing primaryKey?

### DISCUSSION THREAD

**Discussion by jrmypttrsn** [filamentphp/filament#3445] [Help] [answered]
I'm replacing a custom Livewire datatable with Tables and have a view column that contains a button that attaches the model in the datatable to the user's team's watchlist using a custom method in my Livewire. I've been getting random 404 errors so I've been trying to troubleshoot it.

The data I'm using is from an external source so the primaryKey is a string, which I've declared in my model:
```php
protected $primaryKey = 'operation_id';
public $incrementing = false;
protected $keyType = 'string';
```

I'm displaying the primaryKey in my column to ensure that I am indeed getting the correct operation_id. That's all good. 
![Screenshot 2022-08-09 at 13-57-52 myOID](https://user-images.githubusercontent.com/16291625/183742633-e2d6ce84-9f57-466e-8367-025d28a3cbb5.png)

Relevant code for the column:
```php
protected function getTableColumns(): array
{
        return [
            ViewColumn::make('operation_id')
            ->label('')
            ->alignCenter()
            ->view('filament.tables.columns.watchlist-badge'),
            ...
        ]
}
```

The relevant code in ```watchlist-badge.blade.php```:
```html
<x-button.primary wire:click="add({{ $getState() }})" class="has-tooltip p-2 mx-2">
    <x-icon.plus/>
    <span class="tooltip rounded-xl shadow-lg p-2 bg-gray-200 text-sm text-gray-700 -mt-10">
        Add to Team Watchlist
    </span>
</x-button.primary>
<div>
    {{ $getState() }}
</div>
```

My custom method in my Livewire component
```php
public function add(Operation $operation)
{
    dd($operation);
    
    $team = Team::find(auth()->user()->currentTeam->id);
    $team->operations()->syncWithoutDetaching($operation);
    
    $this->notify($operation->operation_name . ' has been added to your team\'s watchlist.');
}
```

What's happening here is that those first four records (0000000000, 0000000001, 0000000008, 0000000009) all attach as expected. Clicking the button for 0000000013 results in a 404.

What is also happening is when I click to the next page (and subsequent paginated records) and click the add button I either get a 404 or the completely wrong record.

Any ideas as to what might be going wrong?

**Comment by danharrin** [comment, marked as answer]
If its a string key, maybe you need to wrap the key in '' instead of outputting it as an integer?

Also instead of $getState() you could use $getRecord()->getKey() which could be clearer as a primary key

**Reply by jrmypttrsn** [reply]
> If its a string key, maybe you need to wrap the key in '' instead of outputting it as an integer?


This was it. I changed
```html
<x-button.primary wire:click="add({{ $getState() }})" class="has-tooltip p-2 mx-2">
```
to 
```html
<x-button.primary wire:click="add('{{ $getRecord()->getKey() }}')" class="has-tooltip p-2 mx-2">
```
and the button works correctly with pagination to boot.

Thank you!

**Reply by danharrin** [reply]
Amazing. Well done

**Reply by jrmypttrsn** [reply]
Haven't done much with non-integer primary keys so this was doing my head in. Thanks again for the clarity and great package.

**Reply by danharrin** [reply]
Let us know if you struggle with anything else 👍
